### PR TITLE
backend: Pass handle token to method calls for easier identification and association with their backend logic

### PR DESCRIPTION
**Pull request by arun-mani-j**
In the current state of `ashpd::backend`, it is difficult to associate each request with the backend logic. So any further calls on the request is left in ambiguity.

For example, let's say two apps call `Account.GetUserInformation`. Our implementation shows dialogs that lets the user enter their details. Now, if one of these apps cancel their request with `Request.close()`, then it is impossible (or difficult) for implementer to know which dialog must be closed.

An easy fix to this problem is to pass the object path of request on the method calls. Now the backend can associate each request with its backend logic, like say a hash map of requests to dialogs. Then, when `close` method is called, it can identify which dialog to close and can close it.

I'm a Rust newbie, so not sure if `clone()`ing paths are a good idea or if we can tackle this problem from a different angle. :sweat_smile: 
Thank you

### CODE REVIEW

File: src/backend/request.rs [hunk 1]
```diff
@@ -39,9 +39,10 @@ impl Request {
         #[cfg(feature = "tracing")]
         tracing::debug!("{_method}");
         let (fut, abort_handle) = abortable(callback);
+        let path_clone = path.clone();
```
Comment (bilelmoussaoui) [inline src/backend/request.rs:42]: I don't really think the path must be passed to every function there is, i am kinda wondering if we should instead pass the AbortHandle...but as it doesn't implement hash, you would not be able to construct a mapping between spawned dialogs and the corresponding requests.

Comment (arun-mani-j) [reply src/backend/request.rs:42]: My reasoning for path is that - it is really simple solution this way and we can get the path from incoming D-Bus request itself without having to store any state in the library side.

Also whatever we are passing for the interface request (like `Account.GetUserInformation`) we need to pass to `Request.close` too so as to establish the association.

Happy to try better ideas and workarounds :)

Comment (bilelmoussaoui) [reply src/backend/request.rs:42]: The thing is I really don't want to expose the Request object path, if we really have to expose part of it, I would expose a HandleToken which we can retrieve from the last portion of the ObjectPath. Basically `HandleToken::try_from(path.as_str().split('/').last().unwrap()).unwrap()` or so.

Where are you using the backend feature btw?

Comment (arun-mani-j) [reply src/backend/request.rs:42]: I think the `HandleToken`s also suit here. Do you have any specific suggestions on how we can expose it to the impls? May be we add a new constructor to `HandleToken` which lets us construct from an `ObjectPath` or so?

We are trying to rewrite Phosh portal in Rust and that's where we are using `ashpd` :).

Comment (bilelmoussaoui) [reply src/backend/request.rs:42]: Interesting! let me know how the experiment goes. 

I would do a TryFrom<ObjectPath> for HandleToken that is gated by the backend feature.

**Comment by arun-mani-j** [comment]
v1:
- Change from using object path to handle token.

File: src/desktop/mod.rs [hunk 2]
```diff
@@ -1,8 +1,8 @@
 mod handle_token;
 pub(crate) mod request;
 mod session;
-pub(crate) use self::handle_token::HandleToken;
 pub use self::{
+    handle_token::HandleToken,
```
Comment (bilelmoussaoui) [inline src/desktop/mod.rs:5]: Should be only exposed if backend is enabled

**Comment by bilelmoussaoui** [comment]
HandleToken also needs to derive various traits like Hash, PartialEq,Eq and so on

File: backend-demo/src/settings.rs [hunk 3]
```diff
@@ -5,7 +5,10 @@ use ashpd::{
         request::RequestImpl,
         settings::{SettingsImpl, SettingsSignalEmitter},
     },
-    desktop::{settings::{ColorScheme, Namespace, APPEARANCE_NAMESPACE, COLOR_SCHEME_KEY}, HandleToken},
+    desktop::{
```
Comment (bilelmoussaoui) [inline backend-demo/src/settings.rs:8]: the formatting fixes should be squashed with the commits that caused them.

File: src/desktop/handle_token.rs [hunk 4]
```diff
@@ -84,6 +87,16 @@ impl TryFrom<&str> for HandleToken {
     }
 }
 
+#[cfg(feature = "backend")]
+impl TryFrom<&OwnedObjectPath> for HandleToken {
+    type Error = HandleInvalidCharacter;
+
+    fn try_from(value: &OwnedObjectPath) -> Result<Self, Self::Error> {
+        let base_segment = value.as_str().split('/').last();
+        HandleToken::try_from(base_segment.unwrap())
```
Comment (bilelmoussaoui) [inline src/desktop/handle_token.rs:96]:
```suggestion
        let base_segment = value.as_str().split('/').last().expect("A valid request ObjectPath");
        HandleToken::try_from(base_segment)
```

File: src/backend/print.rs [hunk 5]
```diff
@@ -53,6 +54,7 @@ impl PrintOptions {
 pub trait PrintImpl: RequestImpl {
     async fn prepare_print(
```
Comment (bilelmoussaoui) [inline src/backend/print.rs:55]: Silence the clippy warning

**Review by bilelmoussaoui — COMMENTED** [review]
otherwise lgtm

File: src/desktop/handle_token.rs [hunk 6]
```diff
@@ -5,7 +5,10 @@ use std::{
 
 use rand::{distributions::Alphanumeric, thread_rng, Rng};
 use serde::{Deserialize, Serialize};
-use zbus::{names::OwnedMemberName, zvariant::Type};
+use zbus::{
+    names::OwnedMemberName,
+    zvariant::{OwnedObjectPath, Type},
```
Comment (bilelmoussaoui) [inline src/desktop/handle_token.rs:10]: This is only usef if the backend feature is enabled, causing a conditional clippy warning

**Comment by arun-mani-j** [comment]
v2:
- HandleToken:
  - Derive Eq, PartialEq and Hash on `backend` feature.
  - Use `.expect` while converting from path.
  - Fix clippy warning on used imports on non-`backend`.
- Desktop:
  - Export `HandleToken` only on `backend` feature.
- Access and Print of Backend:
  - Silence Clippy warning of too many arguments.
- Tree:
  - Perform indent in the same commit.

File: src/backend/access.rs [hunk 7]
```diff
@@ -61,10 +61,12 @@ impl AccessResponse {
     }
 }
 
+#[allow(clippy::too_many_arguments)]
 #[async_trait]
 pub trait AccessImpl: RequestImpl {
     async fn access_dialog(
```
Comment (bilelmoussaoui) [inline src/backend/access.rs:67]:
```suggestion
#[async_trait]
pub trait AccessImpl: RequestImpl {
    #[allow(clippy::too_many_arguments)]
    async fn access_dialog(
```
instead

File: src/backend/print.rs [hunk 8]
```diff
@@ -49,10 +50,12 @@ impl PrintOptions {
     }
 }
 
+#[allow(clippy::too_many_arguments)]
 #[async_trait]
 pub trait PrintImpl: RequestImpl {
     async fn prepare_print(
```
Comment (bilelmoussaoui) [inline src/backend/print.rs:56]:
```suggestion
#[async_trait]
pub trait PrintImpl: RequestImpl {
    #[allow(clippy::too_many_arguments)]
    async fn prepare_print(
```
and here

**Comment by arun-mani-j** [comment]
v3:
- Access and Print of backend:
  - Silence clippy only for the method.

File: src/backend/access.rs (hunk 7 above)
Comment (arun-mani-j) [reply src/backend/access.rs]: That's where I wanted to put too, but by mistake put it over the trait. Thanks.

**Comment by bilelmoussaoui** [comment]
Thanks!
Won't release this anytime soon as it contains breaking API changes. Feel free to use ashpd from git and find what other missing/broken things in ashpd that would make implementing the posh portals easier.

**Comment by arun-mani-j** [comment]
Thank you!
We are iterating the plan for Phosh's new portal in Rust at https://gitlab.gnome.org/arun-mani-j/xdg-desktop-portal-phosh/-/merge_requests/1. 
Thanks again for the support and ashpd :)

**Pull request merged**